src/reactivity.ts

```diff
@@ -103,22 +106,35 @@ function isTrackable(value: any): boolean {
 }
 
 function notifySourceKeyOBservers(source: Source, key: any) {
+  const sourceKeyObservers: Set<Observer> = new Set();
   for (const atom of getSourceAtoms(source)) {
     if (atom[KEYS].has(key)) {
-      atom[OBSERVER]();
+      sourceKeyObservers.add(atom[OBSERVER]);
     }
   }
+
+  for (const observer of getMaximalObservers(sourceKeyObservers)) {
+    observer();
+  }
 }
 
 function notifySourceObservers(source: Source) {
+  const sourceObservers: Set<Observer> = new Set();
   for (const atom of getSourceAtoms(source)) {
-    atom[OBSERVER]();
+    sourceObservers.add(atom[OBSERVER]);
+  }
+
+  for (const observer of getMaximalObservers(sourceObservers)) {
+    observer();
   }
 }
 
-export function registerObserver(observer: Observer) {
+export function registerObserver(observer: Observer, node?: Node) {
   if (!observerAtoms.get(observer)) {
     observerAtoms.set(observer, new Set());
+    node = node || { parent: null };
+    nodeToObserver.set(node, observer);
+    add(node);
```

Comment:
we create a tree with a virtual root that contains all observer nodes and their parents even if they are not observers! 
This simplifies the add operation.